src/types.rs

```diff
@@ -953,10 +953,15 @@ pub fn to_vim_complete_item(
     preferTextEdit: bool,
 ) -> Result<VimCompleteItem> {
     let abbr = lspitem.label.clone();
+    let new_text = match lspitem.text_edit.clone() {
+        Some(text_edit) => text_edit.new_text,
+        None => lspitem.label.clone()
+    };
+
```

Comment:
The haxe language server uses the new "new_text" field for specifying the text.  This doesn't really matter much, since the text edit process in languageclient.rs will fix things.  *I would prefer new_text to be the preferred default, and insert_text to be the fallback, if that's ok I can make that change as well*.

Reply:
`insert_text` is more accurate whenever it is available in situation that `new_text` didn't start from current completion starting point.

Reply:
I think it's ok to use deprecated insert_text when it's available, but you should check for new_text as well rather than immediately falling back to the label.  

Going further, when you get to the "completedone" phase, you can check to see if the completed word is the last edit in the list of text edits to make.  If it is, then you don't need to redo it.  You can just apply the rest of the additionalTextEdits.

Does that make sense?

Reply:
I don't think you got my point here.

`TextEdit`'s `newText` is in a way not compatible with all existing vim completion inforstructure, because the `newText` might not be starting at expected completion starting point by vim.

For example,
```
let language<cursor>
```
where <cursor> is current cursor position.

With vim completion convention (also deoplete's), the returned completion items must all begin with `language`.

Whereas with LSP's `TextEdit`, the `newText` could be `let langugeclient`, with a range start equals to 0.

If in this case, we feed `newText` (`let languageclient`) as `word` to vim, vim completion will behave annoyingly of eating existing input, and in deoplete, it won't be displayed in completion list at all as it doesn't match existing input `language`.

On the other hand, `insertText` and `label` are almost always start with the correct completion starting point.

Reply:
You're right newText is not completely compatible with how vim does completion functions.  However:

1. Most of the time it is, and you can detect it (start/end position are equal to the position the completion was requested at).
2. If it's not, your applyTextEdit method can sort it out, as long as you delete (correctly) the text inserted by the completion engine.

It's true that some of the filtering might be off, at least for some completion engines.  However, that is really a problem for the completion engine, not your plugin.  FWIW deoplete looks like it does the fuzzy matching behavior I would like to have.

Personally, insertText is much simpler for me to deal with.  However, I don't want to try and argue to the haxe languageserver team to support an obsolete field.  FWIW, I've identified some issues on their end where they diverge from spec as well.  I don't personally love the spec, but I really think it's the only way to make sure our code all works together consistently.

Reply:
On a side note, I'll also mention that my rust skills are pretty bad.  I'd hope that you could understand what I'm trying to do and make cleaner/idiomatic changes.  If you see extraneous `to_usize()` methods, please assume I just didn't know what I was doing with low level language features (which is almost certainly true).